app/services/exchange_rates/average_exchange_rates_service.rb

```diff
@@ -0,0 +1,97 @@
+module ExchangeRates
+  class AverageExchangeRatesService
+    VALID_MONTHS = [4, 12].freeze
+    VALID_DAY = 31
+
+    def self.call(force_run:, selected_date:)
+      new(force_run, selected_date).call
+    end
+
+    def initialize(force_run, selected_date)
+      @force_run = force_run
+      @selected_date = Date.parse(selected_date)
+    end
+
+    def call
+      return argument_error unless valid_date || force_run
+
+      avg_rates = create_average_rates
+
+      countries_and_rate = match_country_with_rates(avg_rates)
+      upload_average_rate_file(countries_and_rate)
+    end
+
+    private
+
+    attr_reader :force_run, :selected_date
+
+    def match_country_with_rates(avg_rates)
+      live_countries_last_twelve_months.each_with_object({}) do |country, h|
+        # Need to ensure unique rates are being collected
+        avg_rate = avg_rates
+                    .select { |rate| rate.currency_code == country.currency_code }
+                    .first
+                    .rate
+
+        # { ExchangeRateCountryCurrency => avg_rate }
```

Comment:
do we need this comment?

Reply:
This was just to show what the data structure is tbh

Reply:
Let me know if this comment is ok and adds anything, i only liked it because it shows the structure of the hash better, agreed it was a little more complex when i first did it but happy either way